### PR TITLE
feat: add KV caching support for Wan models

### PR DESCRIPTION
**Pull request by Perseus14**
This Pull Request implements the **KV Cache optimization** for all WAN models (WAN 2.1 & 2.2, both Text-to-Video and Image-to-Video) as well as the **WAN VACE** model.. This optimization pre-computes the Key and Value projections for text and image embeddings before the denoising loop (since they remain constant throughout)

Additionally, this PR introduces a hardware-aware **Dynamic Image Alignment Padding** optimization for next-generation TPUs as well as fix minor bugs and typos.

---

### 1. KV Cache Optimization (WAN & VACE)
* **Attention Level (`FlaxWanAttention`)**: Modified `attention_flax.py` to accept `cached_kv`. If present, key/value projections are bypassed. Added a robust `compute_kv` method to pre-project text (and image) states.
* **Transformer Level (`WanModel` & `WanVACEModel`)**: 
  * Updated `transformer_wan.py` to support KV cache propagation. Added `compute_kv_cache` to precompute block-level cached keys/values.
  * Integrated `skip_embeddings` inside `WanTimeTextImageEmbedding` to bypass redundant embedding layers when using cached states.
  * Extended the same KV cache precomputation and propagation logic to `transformer_wan_vace.py` for both VACE blocks and main blocks.
* **Pipeline Level**: 
  * Updated forward pass helper signatures in `wan_pipeline.py` and `wan_vace_pipeline_2_1.py` to accept and propagate KV caches.
  * Updated all denoising loops (including VACE) to pre-compute the KV cache before starting the loop and reuse it at every step when `use_kv_cache=True`.
* **Config Defaults**: Added `use_kv_cache: False` to all default `.yml` configuration files to ensure backward compatibility.
### 2. Dynamic Image Alignment Padding (Trillium & Ironwood Optimization)
* **Problem**: Image embeddings were previously hardcoded to pad to multiples of `128`. While optimal for older MXU tile sizes (TPU v4, v5p/v5e), next-generation hardware like Trillium (v6e) and Ironwood (v7x) utilize larger **$256 \times 256$** MXU tile structures.
* **Solution**: Replaced hardcoded values with dynamic TPU hardware detection (`get_tpu_type()`). Both `attention_flax.py` and `embeddings_flax.py` (`NNXWanImageEmbedding`) now dynamically adjust image alignment padding:
  * **256-alignment** on Trillium (`v6e`) and Ironwood (`7x`) to perfectly match larger hardware tiles.
  * **128-alignment** fallback on older TPU architectures (`v5p` and below).
### 3. Critical Bug Fixes
* **Fixed `prompt=None` handling in Pipeline**: Updated `encode_prompt` and `__call__` in `wan_pipeline.py` to infer `batch_size` from `prompt_embeds` when `prompt` is `None`, unblocking tests that pass pre-computed embeddings without raw text prompts.
* **Fixed `getattr` with defaults in `pyconfig`**: Changed `__getattr__` in `HyperParameters` class in `pyconfig.py` to raise `AttributeError` instead of `ValueError` when a key is not found. This allows standard Python `getattr(config, "key", default)` to correctly return the default value instead of crashing.
* Fixed deprecated `jax.tree_map` to `jax.tree.map` in `_cast_floating_to`.
* Fixed typo in error message ("T2Vin" -> "T2V in").
---
### Models
* **`attention_flax.py`**:
  * Imported `get_tpu_type` and `TpuType` for dynamic hardware-aware image alignment padding.
  * Integrated `cached_kv` routing inside `FlaxWanAttention.__call__`.
  * Implemented `compute_kv` support for both T2V and I2V cross-attentions.
* **`transformer_wan.py`**:
  * Added `skip_embeddings` parameter inside `WanTimeTextImageEmbedding` to bypass redundant text/image projections.
  * Updated `WanTransformerBlock` and `WanModel` to handle `cached_kv` / `kv_cache` passing.
  * Implemented `WanModel.compute_kv_cache` to precompute block-level cached keys/values across scan and non-scan layers.
* **`transformer_wan_vace.py`**:
  * Added `compute_kv` to `WanVACETransformerBlock`.
  * Added `compute_kv_cache` to `WanVACEModel` to precompute block-level cached keys/values for both VACE and main branches.
  * Updated `__call__` to accept and propagate `kv_cache` to both VACE and main blocks.
* **`modeling_flax_utils.py`**:
  * Fixed deprecated `jax.tree_map` to `jax.tree.map` in `_cast_floating_to`.
* **`embeddings_flax.py`**:
  * Updated `NNXWanImageEmbedding` to dynamically align to `256` for `v6e/7x` and `128` otherwise, avoiding shape mismatches during cross-attention.
### Pipelines
* **`wan_pipeline.py`**:
  * Updated `transformer_forward_pass`, `transformer_forward_pass_full_cfg`, and `transformer_forward_pass_cfg_cache` to accept and pass `kv_cache`.
  * Fixed `NoneType` error when calculating batch size by checking if `prompt` is `None`.
* **`wan_vace_pipeline_2_1.py`**:
  * Added `use_kv_cache` parameter to pipeline calls.
  * Pre-computed `kv_cache` before the denoising loop and passed it to `transformer_forward_pass` in `run_inference`.
* **`wan_pipeline_2_1.py`** & **`wan_pipeline_2_2.py`**:
  * Added `use_kv_cache` parameter to pipeline calls and pre-computed `kv_cache` and `rotary_emb` prior to the denoising loop.
* **`wan_pipeline_i2v_2p1.py`** & **`wan_pipeline_i2v_2p2.py`**:
  * Integrated dynamic pre-computed `kv_cache` support for I2V workflows.
* **`generate_wan.py`**:
  * Fixed typo in error message ("T2Vin" -> "T2V in").

### Configs
* **Configs** (`base_wan_1_3b.yml`, `base_wan_14b.yml`, `base_wan_27b.yml`, `base_wan_i2v_14b.yml`, `base_wan_i2v_27b.yml`):
  * Added `use_kv_cache: False` for all default configs.
  * Modified `prompt` in I2V config files

## Performance Note
* **Observed Latency Savings**: 
  * **~0.2s** on TPU v7x-8 (Ironwood)
  * **~0.5s** on TPU v6e-8 (Trillium)
* **Analysis**: The latency savings during a full denoising run are minimal. This is mathematically expected because the cross-attention Key/Value projections operate on a very small text prompt sequence (typically 512 tokens). The computational FLOPs saved by caching these projections represent a negligible fraction ($< 0.01\%$) of the total workload compared to the massive latent sequence length processed by the self-attention and FFN layers at every step of the denoising loop.

> [!NOTE]
> **Test Configuration**: 720p | 81 frames | 40 steps
> **Hardware**: TPU 7x-8
> **JAX Version**: v0.10.0

| Model | Variant | BaselIne Generation Time | Current Generation Time | Video Link |
|---|---|---|---|---|
| WAN2.2 | T2V | 132.4s | 132.2s | [Link](https://storage.mtls.cloud.google.com/rishabhmanoj-tpu-stuff/maxdiffusion/wan/videos/wan2.2_output_12345_t2v_kv_cache.mp4) |
| WAN2.2 | I2V | 133.6s | 133.3s | [Link](https://storage.mtls.cloud.google.com/rishabhmanoj-tpu-stuff/maxdiffusion/wan/videos/wan2.2_output_12345_i2v_kv_cache.mp4) |
| WAN2.1 | T2V | 132.2s | 132.1s | [Link](https://storage.mtls.cloud.google.com/rishabhmanoj-tpu-stuff/maxdiffusion/wan/videos/wan2.1_output_12345_t2v_kv_cache.mp4) |
| WAN2.1 | I2V | 142.8s | 142.7s | [Link](https://storage.mtls.cloud.google.com/rishabhmanoj-tpu-stuff/maxdiffusion/wan/videos/wan2.1_output_12345_i2v_kv_cache.mp4) |

**Conclusion**: No visual change from baseline across all tested variants.

### CODE REVIEW

File: src/maxdiffusion/pipelines/wan/wan_pipeline_i2v_2p1.py [hunk 1]
```diff
@@ -330,9 +351,25 @@ def run_inference_2_1_i2v(
     image_embeds_combined = image_embeds
```
[the remark this answers is not in the document]
Comment (Perseus14) [reply src/maxdiffusion/pipelines/wan/wan_pipeline_i2v_2p1.py:351]: Done!

File: src/maxdiffusion/pipelines/wan/wan_pipeline_2_2.py [hunk 2]
```diff
@@ -239,6 +254,33 @@ def run_inference_2_2(
   do_classifier_free_guidance = guidance_scale_low > 1.0 or guidance_scale_high > 1.0
```
[the remark this answers is not in the document]
Comment (Perseus14) [reply src/maxdiffusion/pipelines/wan/wan_pipeline_2_2.py:255]: Done!

File: src/maxdiffusion/pipelines/wan/wan_pipeline_2_1.py [hunk 3]
```diff
@@ -267,6 +275,26 @@ def run_inference_2_1(
   cached_noise_cond = None
```
[the remark this answers is not in the document]
Comment (Perseus14) [reply src/maxdiffusion/pipelines/wan/wan_pipeline_2_1.py:275]: Done!

**Comment by prishajain1** [comment]
Has this branch been tested with WAN I2V pipelines to ensure no breakage or performance regressions? (asking as we are making changes to the padding alignment)

**Comment by prishajain1** [comment]
Can we add some unittests with use_kv_cache=True for these pipelines?

File: src/maxdiffusion/models/wan/transformers/transformer_wan_vace.py [hunk 4]
```diff
@@ -467,6 +474,66 @@ def conditional_named_scope(self, name: str):
...
+    if encoder_hidden_states_image is not None:
+      encoder_hidden_states = jnp.concatenate([encoder_hidden_states_image, encoder_hidden_states], axis=1)
+      if encoder_attention_mask is not None:
+        text_mask = jnp.ones(
+            (
+                encoder_hidden_states.shape[0],
+                encoder_hidden_states.shape[1] - encoder_hidden_states_image.shape[1],
+            ),
+            dtype=jnp.int32,
+        )
```
[the remark this answers is not in the document]
Comment (Perseus14) [reply src/maxdiffusion/models/wan/transformers/transformer_wan_vace.py:505]: Done!

File: src/maxdiffusion/models/wan/transformers/transformer_wan.py [hunk 5]
```diff
@@ -652,34 +765,53 @@ def __call__(
...
+            encoder_hidden_states_out,
+            encoder_hidden_states_image_out,
+            encoder_attention_mask_out,
+        ) = self.condition_embedder(
+            timestep,
             encoder_hidden_states,
             encoder_hidden_states_image,
-            encoder_attention_mask,
-        ) = self.condition_embedder(timestep, encoder_hidden_states, encoder_hidden_states_image)
+            skip_embeddings=(kv_cache is not None),
```
[the remark this answers is not in the document]
Comment (Perseus14) [reply src/maxdiffusion/models/wan/transformers/transformer_wan.py:813]: Won't fix as this will cause issues in FlaxWanAttention's slicing logic.

File: src/maxdiffusion/models/attention_flax.py [hunk 6]
```diff
@@ -1591,27 +1626,26 @@ def __call__(
...
         if self.image_seq_len is not None:
           image_seq_len_actual = self.image_seq_len
         else:
           image_seq_len_actual = 257
         padded_img_len = ((image_seq_len_actual + alignment - 1) // alignment) * alignment  # 257 -> 384
-
-        if encoder_attention_mask is None:
-          padded_img_len = image_seq_len_actual
-
         encoder_hidden_states_img = encoder_hidden_states[:, :padded_img_len, :]
```
[the remark this answers is not in the document]
Comment (Perseus14) [reply src/maxdiffusion/models/attention_flax.py:1649]: This is what the commit does

File: src/maxdiffusion/models/embeddings_flax.py [hunk 7]
```diff
@@ -275,7 +276,11 @@ def __init__(
         precision=precision,
     )
     self.norm2 = FP32LayerNorm(rngs=rngs, dim=out_features, elementwise_affine=True, eps=1e-6)
-    self.alignment = alignment
+    if alignment == 128:
+      tpu_type = get_tpu_type()
+      self.alignment = 256 if tpu_type in [TpuType.TPU_V6_LITE, TpuType.TPU_7X] else 128
+    else:
+      self.alignment = alignment
```
[the remark this answers is not in the document]
Comment (Perseus14) [reply src/maxdiffusion/models/embeddings_flax.py]: Done!

File: src/maxdiffusion/models/wan/transformers/transformer_wan_vace.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Perseus14) [reply src/maxdiffusion/models/wan/transformers/transformer_wan_vace.py:505]: Reverted as scan_layers is not implemented for vace

File: src/maxdiffusion/models/modeling_flax_utils.py [hunk 8]
```diff
@@ -93,7 +93,7 @@ def conditional_cast(param):
       return param
 
     if mask is None:
-      return jax.tree_map(conditional_cast, params)
+      return jax.tree.map(conditional_cast, params)
```
[the remark this answers is not in the document]
Comment (Perseus14) [reply src/maxdiffusion/models/modeling_flax_utils.py:96]: Thanks!

File: src/maxdiffusion/pyconfig.py [hunk 9]
```diff
@@ -313,7 +313,7 @@ def __init__(self):
 
   def __getattr__(self, attr):
     if attr not in _config.keys:
-      raise ValueError(f"Requested key {attr}, not in config")
+      raise AttributeError(f"Requested key {attr}, not in config")
```
[the remark this answers is not in the document]
Comment (Perseus14) [reply src/maxdiffusion/pyconfig.py:316]: Thanks!

File: src/maxdiffusion/models/wan/transformers/transformer_wan.py [hunk 10]
```diff
@@ -652,34 +765,87 @@ def __call__(
...
+      # This maintains the correct total sequence length for FlaxWanAttention's internal slicing.
+      # Since these values are ignored when cached_kv is present, the content doesn't matter.
+      if kv_cache is not None and encoder_hidden_states_image_out.shape[-1] != encoder_hidden_states_out.shape[-1]:
+        img_dim = encoder_hidden_states_image_out.shape[-1]
+        text_dim = encoder_hidden_states_out.shape[-1]
+        if img_dim < text_dim:
+          pad_shape = (
+              encoder_hidden_states_image_out.shape[0],
+              encoder_hidden_states_image_out.shape[1],
+              text_dim - img_dim,
```
[the remark this answers is not in the document]
Comment (Perseus14) [reply src/maxdiffusion/models/wan/transformers/transformer_wan.py]: Done!

**Comment by Perseus14** [comment]
@prishajain1 PTAL, added test case for WAN kv_cache and updated the description with inference details

**Comment by Perseus14** [comment]
@ninatu, could you PTAL? I am modifying the WAN Vace files as well.

**Comment by ninatu** [comment]
LGTM

**Review by entrpn — APPROVED** [review]

**Review by prishajain1 — APPROVED** [review]
